File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.java
class Solution {
    public int[] validSequence(String word1, String word2) {
        int n = word1.length();
        int m = word2.length();

        // Memory Fix: Avoid duplicating heavy strings as primitive arrays.
        // We use an int array of size m, which uses minimal memory space.
        int[] lastValidW1Idx = new int[m];
        
        // Inline fill to avoid array mutation overhead
        for (int i = 0; i < m; i++) {
            lastValidW1Idx[i] = -1;
        }

        int j = m - 1;
        for (int i = n - 1; i >= 0; i--) {
            // Using .charAt directly skips memory allocation overhead
            if (j >= 0 && word1.charAt(i) == word2.charAt(j)) {
                lastValidW1Idx[j] = i;
                j--;
            }
        }

        int[] answer = new int[m];
        int w2Idx = 0; 
        boolean changed = false; 

        for (int i = 0; i < n && w2Idx < m; i++) {
            char c1 = word1.charAt(i);
            char c2 = word2.charAt(w2Idx);

            // Case 1: Matching character configuration
            if (c1 == c2) {
                answer[w2Idx] = i;
                w2Idx++;
            } 
            // Case 2: Mutation substitution condition
            else if (!changed && (w2Idx == m - 1 || (lastValidW1Idx[w2Idx + 1] != -1 && lastValidW1Idx[w2Idx + 1] > i))) {
                answer[w2Idx] = i;
                w2Idx++;
                changed = true; 
            }
        }

        if (w2Idx < m) {
            return new int[0];
        }

        return answer;
    }
}

File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {number[]}
 */
var validSequence = function(word1, word2) {
    const n = word1.length;
    const m = word2.length;

    // Memory Optimized: Size m tracking array filled manually to save memory overhead
    const lastValidW1Idx = new Array(m);
    for (let i = 0; i < m; i++) {
        lastValidW1Idx[i] = -1;
    }

    let j = m - 1;
    for (let i = n - 1; i >= 0; i--) {
        // Direct string indexing skips allocation overhead
        if (j >= 0 && word1[i] === word2[j]) {
            lastValidW1Idx[j] = i;
            j--;
        }
    }

    const answer = new Array(m);
    let w2Idx = 0;
    let changed = false;

    for (let i = 0; i < n && w2Idx < m; i++) {
        const c1 = word1[i];
        const c2 = word2[w2Idx];

        // Case 1: Identical match configuration
        if (c1 === c2) {
            answer[w2Idx] = i;
            w2Idx++;
        } 
        // Case 2: Mutation substitution condition
        else if (!changed && (w2Idx === m - 1 || (lastValidW1Idx[w2Idx + 1] !== -1 && lastValidW1Idx[w2Idx + 1] > i))) {
            answer[w2Idx] = i;
            w2Idx++;
            changed = true;
        }
    }

    if (w2Idx < m) {
        return [];
    }

    return answer;
};
